src/track/frame.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+class QDebug;
+
+namespace mixxx {
+class Frame {
```

Comment:
I am struggling with the name `Frame`. It is short and concise, but also wrong.

This represents not an actual *frame of samples*. Instead it is a(n) index/offset/position that addresses a frame i an sequence of frames.

Reply:
@daschuer Suggestions?

Reply:
FramePosition? FrameOffset? FrameIndex?

Reply:
I like FramePosition.

Reply:
We could abbreviate this as `FramePos` to avoid too much type boilerplate.

If we agree to restrict the usage of this class to (absolute) positions the arithmetic operation FramePos + FramePos need to be removed, because adding two points in this affine space is not allowed.

FramePos - FramePos should result in an `double` that implicitly represents the difference (in frames). Otherwise we would need to introduce a dedicated difference (= *vector*) type. I would like to postpone this decision. Let's use a simple type alias for documentation purposes, i.e. `typedef double FrameDiff_t;`. The suffix `_t` indicates that this is a primitive type.

Reply:
So does this mean a frame + frame diff is defined?

Also what about operations that need to find the mid point of two frames? Currently it can be written as (frame1 + frame2) / 2.

Reply:
Maybe add a new method that returns a FramePos and you pass in the other FramePos. This prevents confusing positions with diffs and makes sure we always know what kind of value we're working with.

Reply:
https://en.wikipedia.org/wiki/Affine_space

- points vs. vectors,
- points in time (instants) vs. time durations
- frame positions vs. frame position differences

Reply:
The start of a track can be thought of an origin. Does that mean a frame position is not a true affine space?

Reply:
If I am correct, we have agreed on using only the two types: `FramePos` and `FrameDiff`.
Why do we need to make `FrameDiff`a primitive... can it not be wrapped in a class like `FramePos`?

Reply:
Sort of. Frame positions are relative to the start of the track. Nevertheless adding two frame positions doesn't make any sense.

I assumed that we don't want to define a separate `FrameDiff` type for the moment until we actually need it? That's why I proposed the typedef as a workaround. It depends on how often it appears in public APIs. If those values are only used in internal or limited scopes the separate type is not really needed.

Reply:
This is a snippet from the `Beats` class.
```cpp
void Beats::translate(Frame numFrames) {
    for (BeatList::iterator it = m_beats.begin(); it != m_beats.end();) {
        Frame newpos = Frame(it->frame_position()) + numFrames;
```

It will be changed to 
```cpp
void Beats::translate(FrameDiff numFrames) {
    for (BeatList::iterator it = m_beats.begin(); it != m_beats.end();) {
        FramePos newpos = FramePos(it->frame_position()) + numFrames;
```

Which classes form the public API for mixxx?

Reply:
In Rust I use the newtype pattern very often, because it requires much less boilerplate than in C++. But even then it is a thin line between safety and lots of boilerplate into()/from() type conversions.

Reply:
I would vouch for type safety here.
So I will redefine the operation
```cpp
Frame operator+(const Frame& frame1, const Frame& frame2) {
    return Frame(frame1.getValue() + frame2.getValue());
}
```
as 
```cpp
FramePos operator+(FramePos framePos, FrameDiff frameDiff) {
    return FramePos(framePos.getValue() + frameDiff.getValue());
}
```
If FrameDiff is just a typedef for double, one may still accidentally pass sample diff instead of frame diff.

Reply:
Thinking about this further, do we really need the complexity of creating a class to represent a mathematical operation on a frame position? How about just
```
FramePos someVariable = whatever;
someVariable = FramePos(someVariable.getValue() + increment);
```

Reply:
I agree with @Be. Introducing new types is a lot of extra work. And even more if you want to cover both integer and floating-point types. Out of scope of this task.

Let's instead use naming conventions (as suggested) for variables and parameters to clearly distinguish between sample and frame deltas.

Reply:
Okay, so let's remove the math operators for the Frame class.

Reply:
Thanks for pointing the issues with the abstraction. The code makes much more sense with `FrameDiff_t` type.